File: sdk/python/maestro_sdk/qset.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional

import httpx


@dataclass
class CreateRequestPayload:
    requester: str
    tool: str
    purpose: str
    scopes: Optional[list[str]] = None
    expires_at: Optional[str] = None

    def to_json(self) -> Dict[str, Any]:
        payload: Dict[str, Any] = {
            "requester": self.requester,
            "tool": self.tool,
            "purpose": self.purpose,
        }
        if self.scopes is not None:
            payload["scopes"] = self.scopes
        if self.expires_at is not None:
            payload["expiresAt"] = self.expires_at
        return payload


class QsetClient:
    """Minimal client for the QSET service."""

    def __init__(self, base_url: str, approver_key: Optional[str] = None, *, client: Optional[httpx.Client] = None) -> None:
        self._base_url = base_url.rstrip('/')
        self._approver_key = approver_key
        self._client = client or httpx.Client(base_url=self._base_url)

    def close(self) -> None:
        self._client.close()

    def create_request(self, payload: CreateRequestPayload | Dict[str, Any]) -> Dict[str, Any]:
        body = payload.to_json() if isinstance(payload, CreateRequestPayload) else payload
        return self._post('/requests', json=body)

    def get_request(self, request_id: str) -> Dict[str, Any]:
        return self._get(f'/requests/{request_id}')

    def approve_request(self, request_id: str, approver_key: Optional[str] = None) -> Dict[str, Any]:
        return self._post(f'/requests/{request_id}/approve', approver_key=approver_key)

    def deny_request(self, request_id: str, approver_key: Optional[str] = None) -> Dict[str, Any]:
        return self._post(f'/requests/{request_id}/deny', approver_key=approver_key)

    def mint_token(self, request_id: str, approver_key: Optional[str] = None) -> Dict[str, Any]:
        return self._post(f'/requests/{request_id}/mint', approver_key=approver_key)

    def attenuate_token(self, token_id: str, payload: Dict[str, Any], approver_key: Optional[str] = None) -> Dict[str, Any]:
        return self._post(f'/tokens/{token_id}/attenuate', json=payload, approver_key=approver_key)

    def get_ledger_public_key(self) -> str:
        response = self._get('/ledger/public-key')
        return response['publicKey']

    def _headers(self, approver_key: Optional[str] = None) -> Dict[str, str]:
        headers = {'Content-Type': 'application/json'}
        key = approver_key or self._approver_key
        if key:
            headers['X-Approver-Key'] = key
        return headers

    def _get(self, path: str) -> Dict[str, Any]:
        response = self._client.get(path, headers=self._headers())
        response.raise_for_status()
        return response.json()

    def _post(
        self,
        path: str,
        *,
        json: Optional[Dict[str, Any]] = None,
        approver_key: Optional[str] = None,
    ) -> Dict[str, Any]:
        response = self._client.post(path, json=json, headers=self._headers(approver_key))
        response.raise_for_status()
        return response.json()

    def __enter__(self) -> "QsetClient":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

File: sdk/typescript/src/qset.ts
export interface CreateRequestPayload {
  requester: string;
  tool: string;
  purpose: string;
  scopes?: string[];
  expiresAt?: string;
}

export interface RequestRecord {
  id: string;
  requester: string;
  tool: string;
  purpose: string;
  scopes: string[];
  expiresAt: string;
  createdAt: string;
  status: 'pending' | 'approved' | 'denied';
  approvals: Record<string, string>;
  denials: Record<string, string>;
  tokenId?: string;
}

export interface TokenRecord {
  id: string;
  requestId: string;
  parentId?: string;
  scopes: string[];
  expiresAt: string;
  issuedAt: string;
  secret: string;
}

export interface AttenuatePayload {
  scopes?: string[];
  expiresAt?: string;
}

export class QsetClient {
  constructor(private readonly baseUrl: string, private readonly approverKey?: string) {}

  async createRequest(payload: CreateRequestPayload): Promise<RequestRecord> {
    return this.post<RequestRecord>('/requests', payload);
  }

  async getRequest(id: string): Promise<RequestRecord> {
    return this.get<RequestRecord>(`/requests/${id}`);
  }

  async approveRequest(id: string, approverKey?: string): Promise<RequestRecord> {
    return this.post<RequestRecord>(`/requests/${id}/approve`, undefined, approverKey);
  }

  async denyRequest(id: string, approverKey?: string): Promise<RequestRecord> {
    return this.post<RequestRecord>(`/requests/${id}/deny`, undefined, approverKey);
  }

  async mintToken(id: string, approverKey?: string): Promise<TokenRecord> {
    return this.post<TokenRecord>(`/requests/${id}/mint`, undefined, approverKey);
  }

  async attenuateToken(id: string, payload: AttenuatePayload, approverKey?: string): Promise<TokenRecord> {
    return this.post<TokenRecord>(`/tokens/${id}/attenuate`, payload, approverKey);
  }

  async getLedgerPublicKey(): Promise<string> {
    const response = await this.get<{ publicKey: string }>('/ledger/public-key');
    return response.publicKey;
  }

  private async get<T>(path: string): Promise<T> {
    const response = await fetch(this.baseUrl + path, {
      headers: this.headers(),
    });
    if (!response.ok) {
      throw this.error(response);
    }
    return (await response.json()) as T;
  }

  private async post<T>(path: string, body?: unknown, approverKey?: string): Promise<T> {
    const headers = this.headers(approverKey);
    const response = await fetch(this.baseUrl + path, {
      method: 'POST',
      headers,
      body: body === undefined ? undefined : JSON.stringify(body),
    });
    if (!response.ok) {
      throw this.error(response);
    }
    return (await response.json()) as T;
  }

  private headers(approverKey?: string): HeadersInit {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    const key = approverKey ?? this.approverKey;
    if (key) {
      headers['X-Approver-Key'] = key;
    }
    return headers;
  }

  private error(response: Response): Error {
    return new Error(`QSET request failed: ${response.status}`);
  }
}
